File: 第二章-线性表、栈和队列/代码/Python,C/线性表.py
"""
@File:    线性表.py
@Author:  csh
@Date:    2020/05/21
"""


# 线性表的顺序存储可以用列表进行


# 线性表的链式存储
class Node(object):
    """节点"""
    def __init__(self, value):
        """初始化一个节点

        :param value 节点元素
        """
        self.__data = value
        self.__next = None

    def setData(self, newValue):
        """设置当前节点元素

        :param newValue 节点元素
        """
        self.__data = newValue

    def getData(self):
        """获取当前节点的元素值

        :return data 节点元素
        """
        return self.__data

    def setNext(self, nextNode):
        """设置下一个节点对象

        :param nextNode Node对象
        """
        self.__next = nextNode

    def getNext(self):
        """获取当前节点的下一个节点对象

        :return Node对象
        """
        return self.__next


class UnOrderedList(object):
    """无序链表"""
    def __init__(self):
        """初始化一个头尾节点"""
        self.__head = None

    def setPos(self, p):
        """返回第p个节点对象

        :return Node对象
        """
        count = 1;
        if p <= 0:
            return None
        temp = self.__head
        while count < p and temp != None:
            count += 1
            temp = temp.getNext()
        return temp

    def isEmpty(self):
        """判断链表是否为空

        return bool
        """
        return self.__head == None

    def length(self):
        """链表的长度

        :return int
        """
        temp = self.__head
        count = 0
        while temp != None:
            count += 1
            temp = temp.getNext()
        return count

    def append(self, value):
        """添加一个元素

        :param value 节点元素
        """
        temp = Node(value)
        temp.setNext(self.__head)
        self.__head = temp

    def insert(self, num, item):
        """第num个位置插入节点元素

        :param num int
        :param item Node对象
        """
        if num <= 0 or num > self.length()+1:
            print("Insertion point is illegal")
            return
        previous = self.setPos(num-1)
        temp = Node(item)
        if previous == None:
            # 此时num-1=0，即插入到头节点
            temp.setNext(self.__head)
            self.__head = temp
        else:
            current = previous.getNext()
            temp.setNext(current)
            previous.setNext(temp)

    def remove(self, p):
        """移除位置为p的元素

        :param p int
        """
        if p <= 0 or p > self.length():
            print("Removal point is illegal")
            return
        if p == 1:
            self.__head = self.__head.getNext()
        else:
            previous = self.setPos(p-1)
            current = previous.getNext()
            previous.setNext(current.getNext())

    def getPos(self, value):
        """查找value元素是否存在

        :param value 元素值
        :return bool
        """
        current = self.__head
        found = False
        while current != None and not found:
            if current.getData() == value:
                found = True
            else:
                current = current.getNext()
        return found

    def getValue(self, p):
        """获取位置为p的元素值

        :param p int
        :return data 元素值
        """
        current = self.setPos(p)
        if current == None:
            print("Location is illegal")
        return current.getData()

    def showAll(self):
        """显示所有元素值
        """
        tmp = self.__head
        while tmp != None:
            print(tmp.getData())
            tmp = tmp.getNext()


class OrderedList(object):
    """有序链表"""
    def __init__(self):
        """初始化头节点"""
        self.__head = None

    def isEmpty(self):
        """判断链表是否为空

        :return bool
        """
        return self.__head == None

    def length(self):
        """链表的长度
        
        :return int
        """
        count = 0
        temp = self.__head
        while temp != None:
            count += 1
            temp = temp.getNext()
        return count

    def append(self, value):
        """添加一个元素

        :param value 元素值
        """
        current = self.__head
        previous = None
        stop = False
        while current != None and not stop:
            if current.getData() > value:
                stop = True
            else:
                previous = current
                current = current.getNext()
        temp = Node(value)
        if previous == None:
            # 当前没有元素
            temp.setNext(self.__head)
            self.__head = temp
        else:
            temp.setNext(current)
            previous.setNext(temp)

    def remove(self, item):
        """移除原始值为item的节点

        :param item 要移除的节点的元素值
        """
        current = self.__head
        previous = None
        found = False
        while not found:
            if current.getData() == item:
                found = True
            else:
                previous = current
                current = current.getNext()
        if previous == None:
            self.__head = current.getNext()
        else:
            previous.setNext(current.getNext())

    def find(self, value):
        """查找元素值是否存在

        :param value 节点元素值
        :return bool
        """
        current = self.__head
        found = False
        stop = False
        while current != None and not found and not stop:
            if current.getData() == value:
                found = True
            else:
                if current.getData() > value:
                    stop = True
                else:
                    current = current.getNext()
        return found

    def showAll(self):
        current = self.__head
        while current != None:
            print(current.getData(), end=" ")
            current = current.getNext()
        print()

File: 第二章-线性表、栈和队列/代码/Python,C/线性表.c
// File:    线性表.c
// Author:  csh
// Date:    2020/05/23
// ===================


#include <stdlib.h>
#include <stdio.h>

#define TRUE 1
#define FALSE 0
#define OK 1

typedef int bool;


// ************************** 线性表的顺序存储 *********************************//
#define MAXSIZE 10000
typedef int T;

typedef struct Node *List;
struct Node{
    T Data[MAXSIZE];
    int length;
};

List InitList();                                            // 创建一个空表
int lenList(List Ptrl);                                     // 线性表大小
bool clearList(List Ptrl);                                  // 清空线性表
bool isEmptyList(List Ptrl);                                // 判断线性表是否为空
bool isFullList(List Ptrl);                                 // 判断线性表是否已满
bool appendList(List Ptrl, const T value);                  // 表尾添加元素
bool insertList(List Ptrl, const int p, const T value);     // 第P个位置插入元素value
bool removeList(List Ptrl, const int p);                    // 移除第p个元素
bool setValueList(List Ptrl, const int p, const T value);   // 设置第p个位置的元素
bool getValueList(List Ptrl, const int p, T *value);        // 获取第p个元素
bool getPosList(List Ptrl, const T value, int *p);          // 获取元素为value的位置
void showAllList(List Ptrl);

List InitList(){
    List Ptrl;
    Ptrl = malloc(sizeof(struct Node));
    if(!Ptrl) exit(-1);
    Ptrl->length = 0;
    return Ptrl;
}

int lenList(List Ptrl){
    return Ptrl->length;
}

bool clearList(List Ptrl){
    Ptrl->length = 0;
    return TRUE;
}

bool isEmptyList(List Ptrl){
    return Ptrl->length == 0;
}

bool isFullList(List Ptrl){
    return Ptrl->length == MAXSIZE;
}

bool appendList(List Ptrl, const T value){
    if(Ptrl->length >= MAXSIZE){
        printf("OVERFLOW!");
        return FALSE;
    }
    Ptrl->Data[(Ptrl->length)++] = value;
    return TRUE;
}

bool insertList(List Ptrl, const int p, const T value){
    if(p < 1 || p > Ptrl->length+1){
        printf("Insertion point is illegal");
        return FALSE;
    }
    if(Ptrl->length == MAXSIZE){
        printf("List Full");
        return FALSE;
    }
    for(int j = Ptrl->length-1; j >= p-1; j--)
        Ptrl->Data[j+1] = Ptrl->Data[j];
    Ptrl->Data[p-1] = value;
    Ptrl->length ++;
    return TRUE;
}

bool removeList(List Ptrl, const int p){
    if(p < 1|| p > Ptrl->length){
        printf("Removal point is illegal");
        return FALSE;
    }
    for(int j = p; j <= Ptrl->length-1; j++)
        Ptrl->Data[j-1] = Ptrl->Data[j];
    Ptrl->length --;
    return TRUE;
}

bool getValueList(List Ptrl, const int p, T *value){
    if(p < 1 || p > Ptrl->length){
        printf("Location is illegal");
        return FALSE;
    }
    *value = Ptrl->Data[p-1];
    return TRUE;
}

void showAllList(List Ptrl){
    for(int i = 0; i < Ptrl->length; i++)
        printf("%d ", Ptrl->Data[i]);
    printf("\n");
}

bool getPosList(List Ptrl, const T value, int *p){
    int i = 0;
    while(i < Ptrl->length && Ptrl->Data[i] != value)
        i++;
    if(i >= Ptrl->length)
        return FALSE;
    *p = i+1;
    return TRUE;
}

// *********************************************************************//


// ************************** 线性表的链式存储 *********************************//

   /*
   这里和C++写法不同，没有定义尾指针，所以添加元素等操作都有所不同，显示所有
   元素时可发现是逆序显示
   */

typedef struct LinkNode* Link;

struct LinkNode{
    T data;
    Link next;
};

Link InitLink();                                            // 创建一个空表
int lenLink(Link Ptrl);                                     // 线性表大小
Link setPos(const int p, Link Ptrl);                        // 找到第p个位置的结点
bool clearLink(Link Ptrl);                                  // 清空线性表
bool isEmptyLink(Link Ptrl);                                // 判断线性表是否为空
bool appendLink(Link Ptrl, const T value);                  // 表尾添加元素
bool insertLink(Link Ptrl, const int p, const T value);     // 第P个位置插入元素value
bool removeLink(Link Ptrl, const int p);                    // 移除第p个元素
bool setValueLink(Link Ptrl, const int p, const T value);   // 设置第p个位置的元素
bool getValueLink(Link Ptrl, const int p, T *value);        // 获取第p个元素
bool getPosLink(Link Ptrl, const T value, int *p);          // 获取元素为value的位置
void showAllLink(Link Ptrl);

// *********************************************************************//

int main(){
    return 0;
}
